### ISSUE TITLE
Troubleshooting piping in hw2

### ISSUE THREAD

**Issue by jduga002** [mikeizbicki/ucr-cs100#1002]
I'm getting a bug in my program when I try piping. For some reason, the second program in the pipe will not end. For example, if I am in rshell and I run

```
$  ls | grep i
bin
existingInputFile
Makefile
                         #sits here doing nothing
```

is the output I get, and it just hangs on the second command waiting for it to finish. So you can see it pipes the input correctly to grep, but then grep never finishes execution.

Using gdb, I've found the line of code where it hangs is when I call waitpid() a second time.  Here is the relevant code:

```
// v_commands holds the commands to be executed
// for example if I entered the command in the example above,
// v_commands would hold two vectors of char pointers:
// {  {"ls", NULL}, {"grep", "i", NULL} }
int exec_commands_iopip(vector<vector<char *> > &v_commands, const vector<int *> &v_fds) {
    vector<int> v_pids;
    for (unsigned i = 0; i < v_commands.size(); i++) {
        if (i != v_commands.size()-1) {
            if (-1 == pipe(v_fds.at(i))) {
                perror("Error with pipe");
                return 1;
            }
        }
        int pid = fork();
        if (pid == -1) { //error with fork
            perror("error with fork");
            return 1;
        }
        else if (pid == 0) { //child process
            if (i != v_commands.size()-1) {
                if (-1 == dup2(v_fds.at(i)[P_WRITE],1)) {
                    perror("Error with dup2");
                    exit(1);
                }
            }
            if (i != 0) {
                if (-1 == dup2(v_fds.at(i-1)[P_READ],0)) {
                    perror("Error with close");
                    exit(1);
                }
            }
            char **command_arr = &v_commands.at(i)[0];
            if (-1 == execvp(command_arr[0], command_arr)) {
                perror("execvp");
                exit(1);
            }
            exit(0);
        }
        else { //parent process
            v_pids.push_back(pid);
            //continue loop
        }
    }
    for (unsigned i = 0; i < v_commands.size(); i++) {
        int x;
        if (-1 == waitpid(v_pids.at(i), &x, 0)) { // program hangs here when i == 1
            perror("wait"); 
            exit(1);
        }
        if (i != 0) {
            if (-1 == close(v_fds.at(i-1)[P_READ])) {
                perror("close");
                exit(1);
            }
            if (-1 == close(v_fds.at(i-1)[P_WRITE])) {
                perror("close");
                exit(1);
            }
        }
    }
    return 0;
}
```

**Comment by mikeizbicki** [comment]
This is a good question.  It clearly describes the problem, the attempts to fix it, and why those fixes are not working.  +3 pts extra credit.

That said, you still haven't _proven_ to me that "the second program in the pipe will not end" like you claim.  Until we know that's actually the problem, we can't look for a solution.  Try opening up another terminal and running `ps -ef | grep grep` in bash.  This will list all the processes and print them out if their name includes the string `grep`.  If you see the process listed, then you know the process isn't terminating like you think and we can begin fixing that problem.  If you don't see the process listed, then you know something else is going on and we have to figure out what the actual problem is.  

The output might be less confusing if you use a different command than grep for the test case.

**Comment by jduga002** [comment]
Okay, so when `rshell` is not running, and I run `ps -ef | grep grep`, I get the result

```
dr1zz1t+  7191     1  0 16:37 pts/5    00:00:00 grep .
dr1zz1t+ 15595 15571  0 18:34 pts/6    00:00:00 grep --colour=auto grep
```

But after opening `rshell` and typing in `ls | grep i` and letting it run for a couple minutes, then `ps -ef | grep grep` returns

```
dr1zz1t+  7191     1  0 16:37 pts/5    00:00:00 grep .
dr1zz1t+ 15599 15596  0 18:35 pts/5    00:00:00 grep i
dr1zz1t+ 15852 15571  0 18:39 pts/6    00:00:00 grep --colour=auto grep
```

**Comment by mikeizbicki** [comment]
Is there a particular reason you're using `waitpid` instead of `wait`?  The next thing I would check is to make sure you're actually waiting on the right process.

**Comment by jduga002** [comment]
I had originally used `wait`, but I was having this problem, so I switched to `waitpid` to try to fix it, which it didn't. I've changed it back to `wait` now.

**Comment by mikeizbicki** [comment]
So now we've determined that `grep` the problem is actually that grep is not terminating and we're definitely waiting on the right program.  The next step is to think, what are the reasons grep wouldn't terminate?  It keeps running until it doesn't receive any more input from stdin.  So if stdin never gets closed, then grep will keep running.  So how can we verify that stdin is actually getting closed?

**Comment by jduga002** [comment]
So after running `rshell`, I opened a new terminal and ran the following:

```
$ ps -ef | grep rshell
dr1zz1t+  2559  2404  0 10:54 pts/2    00:00:00 bin/rshell
dr1zz1t+  2561  2449  0 10:54 pts/3    00:00:00 grep --colour=auto rshell
$ ls /proc/2559/fd
0  1  2
```

Then in my terminal running `rshell` I typed `ls | grep i`. Then back in my second terminal, I ran

```
$ ls /proc/2559/fd
0  1  2  3  4
```

So now I can see I have some extra file descriptors still open. So I `cd`'d into `/proc/2559/fd` to try `echo`ing some phrases into my `rshell` terminal. Here is what I ran and whether it succeeded or not:

`echo hello wirld > 0`        succeeded
`echo hello world > 0`        succeeded
`echo hello world > 3`        nothing printed
`echo hello wirld > 3`        succeeded

So it looks like file descriptor `3` is `stdin` for `grep` and is not closing, as strings passing through `3` are being filtered for the letter `i`. So now I need to figure out where to move my `close` syscalls to make it work.

Moving around the last part like this:

```
for (unsigned i = 0; i < v_commands.size(); i++) {
     int x;
     if (-1 == wait(&x)) { // parent function waits for child
         perror("wait");  // program exits if there is an error
         exit(1);
     }
     if (i < v_commands.size()-1) {
         if (-1 == close(v_fds.at(i)[P_WRITE])) {
             perror("close");
             exit(1);
         }
         if (-1 == close(v_fds.at(i)[P_READ])) {
             perror("close");
             exit(1);
         }
     }
 }
```

causes the file descriptors in the corresponding folder in `/proc` to be just `0 1 2`,  but the program still hangs. Should I be manually closing stdin and stdout near the beginning and then restoring it at the end?

**Comment by mikeizbicki** [comment]
+3 more points extra credit for the detailed and insightful follow up.

My guess was that changing `i-1` to `i` like you did was going to fix everything.  That was definitely a bug, but it looks like there's another bug lurking here somewhere.

We said that grep will terminate when stdin closes, and we're pretty sure that stdin is closing now.  So there is probably something else causing the second program in the pipe to keep running.  A good way to make sure this has nothing to do with stdin is to run a program that doesn't use stdin at all.  For example, `ls | ls`.  

It would also be worth doing some tests with multiple pipes.  When you run `ls | grep i | grep M` do you get the expected results?  Do both greps fail to close or only the last one?

**Comment by jduga002** [comment]
I had already tried this last night, but since I've modified the program a little this morning I tried it again.  And I got the same results.

`ls | ls` works just like it should. I also tried combinations of `ls` and `ls -a` to make sure it was only printing the second one's stdout to the terminal, which it is. Now `ls | grep i | grep M` freezes with no output, and gdb shows that it successfully waits for the first process, then closes the first pipe, and then gets stuck waiting for the second process. Also, after running `ls | grep i | grep M`, the listed file descriptors in the corresponding folder in `/proc` are `0 1 2 3 4`. So I tried doing output redirection into these file descriptors, and whatever word I used `echo` on, in only printed in the `rshell` terminal if it was redirected into `0`, `1`, or `2`.

Then I decided to try something else, and this is really interesting. After running `rshell`, I did this:

```
$ ls | ls -a | grep M
close: Bad file descriptor
dr1zz1tdudden@dr1zz1tdudden-VirtualBox ~/Documents/rshell $ Makefile
README.md

```

This all just happened automatically after typing in  `ls | ls -a | grep M` Bash then sat there waiting for me to type something, so I pressed enter and it showed the command prompt for bash. The interesting thing is, that running `ps -ef | grep grep` shows that `grep M` is still running, even though `rshell` closed. Trying it again with `gdb` shows that `perror()` is called after trying to close the file descriptor the second time through the loop, i.e. at `close(v_fds.at(1)[P_WRITE])`.

So for some reason, the file descriptors for the second pipe are bad, so the program exits, but the output of `ls -a` is still piped to `grep M`, and `grep M`'s output is printed to stdout after `rshell` closes, though `grep M` is still running.

Lastly, in rshell I ran:

```
$ ls | ls -l | ls -a
.  ..  bin  existingInputFile  .git  grep  LICENSE  Makefile  README.md  src  tests
close: Bad file descriptor
```

so this also terminated the program because of a bad file descriptor; however, `ps -ef | grep ls` shows that `ls` is not still running in the background.

**Comment by jduga002** [comment]
So after a bit of searching the web, I found [this article](http://stackoverflow.com/questions/7281894/how-do-i-chain-stdout-in-one-child-process-to-stdin-in-another-child-in-c/7282296#7282296) that was able to fix the problem. I changed from using a vector of (n-1) int arrays to just two int arrays, and then I added close statements to all the parts the article suggested, and now my piping works!

 So the important point to take away from this is to always close your unused file descriptors in _both_ the parent _and_ the child.
